File: ios/SpatailEducator/Sources/RoomModel.swift
import Foundation
import simd

// RoomModel — the SPATAIL placement solver's INPUT on device (mirrors
// studio/spatail/room_model.py). A compact semantic description of the user's
// space, built on-device from ARKit and consumed by PlacementSolver. NOT a mesh.
//
// Frame: metres, +Y up, user near origin looking -Z (the AR convention).

struct RoomModel {
    struct Surface {
        var cls: String                 // floor | table | wall | ceiling | seat | unknown
        var center: SIMD3<Float>
        var size: SIMD2<Float>          // extent along the surface (w, d)
        var normal: SIMD3<Float>
        var yaw: Float
        var height: Float               // height above floor (table ≈ 0.74)
        var confidence: Float
    }
    struct Obstacle {
        var bboxMin: SIMD3<Float>
        var bboxMax: SIMD3<Float>
        var category: String
    }
    struct Light {
        var intensity: Float            // ARLightEstimate.ambientIntensity (lumens)
        var colorTempK: Float
        var keyDirection: SIMD3<Float>
    }

    var surfaces: [Surface] = []
    var obstacles: [Obstacle] = []
    var user = (position: SIMD3<Float>(repeating: 0), forward: SIMD3<Float>(0, 0, -1), eyeHeight: Float(1.45))
    var light = Light(intensity: 1000, colorTempK: 6500, keyDirection: [0, -1, 0])
    var bounds = SIMD3<Float>(4, 3, 2.6)
    var source = "default"

    func surfaces(of cls: String) -> [Surface] { surfaces.filter { $0.cls == cls } }
    /// The table the USER is engaging with — not the biggest slab in the room.
    /// Gate out implausible "tables" first (ARKit's ML sometimes labels floor-level
    /// planes .table), then prefer near + in-front of the user (§6: centre on the
    /// detected table, main content in front of the user); area is only a tiebreak.
    func bestTable() -> Surface? {
        let floorH = floor()?.height
        let plausible = surfaces(of: "table").filter { t in
            let above = floorH.map { t.height - $0 } ?? 0.7   // no floor yet → trust the label
            return above > 0.25 && above < 1.3 && t.size.x * t.size.y > 0.05
        }
        func score(_ t: Surface) -> Float {
            let to = SIMD3(t.center.x - user.position.x, 0, t.center.z - user.position.z)
            let d = simd_length(to)
            let behind: Float = d > 0.3 && simd_dot(to / d, user.forward) < 0 ? 2.0 : 0
            return d + behind - min(t.size.x * t.size.y, 1.0) * 0.3
        }
        return plausible.min { score($0) < score($1) }
    }
    func floor() -> Surface? {
        surfaces(of: "floor").max { $0.size.x * $0.size.y < $1.size.x * $1.size.y }
    }
}

#if os(iOS)
import ARKit

// Accumulates ARKit readings over the session and distills a RoomModel — the
// on-device twin of analysis.py's plane→RoomProfile path, upgraded to carry
// semantic plane classification + light. (Raw LiDAR-mesh occlusion/obstacles are
// a later refinement; planes + classification already give a solid model.)
@MainActor
final class RoomModelBuilder {
    private var planes: [UUID: ARPlaneAnchor] = [:]
    private var light = RoomModel.Light(intensity: 1000, colorTempK: 6500, keyDirection: [0, -1, 0])
    private var userPos = SIMD3<Float>(repeating: 0)
    private var userFwd = SIMD3<Float>(0, 0, -1)

    func update(planes anchors: [ARPlaneAnchor]) { for a in anchors { planes[a.identifier] = a } }
    func remove(_ ids: [UUID]) { for id in ids { planes[id] = nil } }

    func update(light est: ARLightEstimate?) {
        guard let e = est else { return }
        light = RoomModel.Light(intensity: Float(e.ambientIntensity),
                                colorTempK: Float(e.ambientColorTemperature),
                                keyDirection: [0, -1, 0])
    }

    func update(cameraTransform t: simd_float4x4) {
        userPos = SIMD3(t.columns.3.x, t.columns.3.y, t.columns.3.z)
        let f = -SIMD3(t.columns.2.x, t.columns.2.y, t.columns.2.z)   // -Z is forward
        let flat = SIMD3(f.x, 0, f.z)
        userFwd = simd_length(flat) > 1e-4 ? simd_normalize(flat) : SIMD3(0, 0, -1)
    }

    private static func classify(_ p: ARPlaneAnchor, floorY: Float, cameraY: Float) -> String {
        switch p.classification {
        case .floor:   return "floor"
        case .table:   return "table"
        case .wall:    return "wall"
        case .ceiling: return "ceiling"
        case .seat:    return "seat"
        default:
            // unclassified: infer from height ABOVE THE FLOOR. ARKit world y=0 is the
            // session-start phone pose (~1.4 m up), so an absolute threshold labels
            // every table "floor" and the solver shoves content past the real table.
            if p.alignment == .horizontal {
                let y = p.transform.columns.3.y
                if y > cameraY + 0.5 { return "ceiling" }
                return y - floorY < 0.35 ? "floor" : "table"
            }
            return "wall"
        }
    }

    func build() -> RoomModel {
        var rm = RoomModel()
        rm.source = "arkit"
        rm.light = light
        rm.user = (position: userPos, forward: userFwd,
                   eyeHeight: max(Float(1.2), userPos.y > 0.1 ? userPos.y : Float(1.45)))
        var minX: Float = 0, maxX: Float = 0, minZ: Float = 0, maxZ: Float = 0
        // floor reference for the height heuristic: the lowest horizontal plane seen,
        // or the camera's typical handheld height above the floor when none is lower
        // (a lone table plane must not become its own "floor")
        let lowestHorizontal = planes.values
            .filter { $0.alignment == .horizontal }
            .map { $0.transform.columns.3.y }.min() ?? .greatestFiniteMagnitude
        let floorY = lowestHorizontal <= userPos.y - 1.0 ? lowestHorizontal : userPos.y - 1.4
        for (_, p) in planes {
            let c = p.transform.columns.3
            let ext = p.planeExtent
            rm.surfaces.append(RoomModel.Surface(
                cls: Self.classify(p, floorY: floorY, cameraY: userPos.y),
                center: SIMD3(c.x, c.y, c.z),
                size: SIMD2(ext.width, ext.height),
                normal: SIMD3(0, 1, 0),
                yaw: ext.rotationOnYAxis,
                height: c.y,
                confidence: 1))
            minX = min(minX, c.x - ext.width / 2);  maxX = max(maxX, c.x + ext.width / 2)
            minZ = min(minZ, c.z - ext.height / 2); maxZ = max(maxZ, c.z + ext.height / 2)
        }
        rm.bounds = SIMD3(max(4, maxX - minX), max(3, maxZ - minZ), 2.6)
        return rm
    }
}
#endif

File: studio/spatail/room_model.py
"""room_model.py — the SPATAIL RoomModel: the placement solver's INPUT.

RoomModel is a compact, SEMANTIC description of the user's physical space — "the
room, understood." It is the input to the Placement Design System solver. It is
NOT a mesh or a scan file; it is the small structured abstraction the placement
rules reason over, so the same rules run identically on device (live) and on PC
(authoring/preview/tests).

WHERE IT COMES FROM (all on-device, accumulated over a few seconds of looking
around, then distilled — see ios .../ARContainerView.swift, which already builds
the coarse RoomProfile this evolves from):

    ARKit source                              -> RoomModel field
    ----------------------------------------- -------------------------------------
    ARPlaneAnchor (.planeExtent/.transform/   -> surfaces[]  (with .classification:
        .alignment/.classification)              floor/wall/ceiling/table/seat/...)
    ARMeshAnchor (sceneReconstruction =       -> surfaces[] (precise), obstacles[],
        .meshWithClassification, LiDAR)          occlusion geometry
    ARFrame.lightEstimate                     -> light (intensity, colorTemp)
    ARFrame.camera.transform                  -> user (position, forward, eyeHeight)
    ARFrame.sceneDepth (LiDAR)                -> obstacles / free-space refinement
    RoomPlan CapturedRoom (optional one-shot) -> a near-complete RoomModel
    Vision/CoreML object detection (optional) -> obstacles[] with object identity

Producers (ARKit / RoomPlan / CV) all WRITE this schema; the solver only ever
READS it — so producers are swappable without touching placement logic.

Frame: metres, +Y up, the user near the origin looking -Z (the AR/USDZ convention
the device uses; matches studio/spatail/analysis.py).
"""
from __future__ import annotations

from dataclasses import dataclass, field, asdict

SCHEMA = "spatail-room-model/1"

# Semantic surface classes (mirror ARKit's ARMeshClassification / plane classes).
SURFACE_CLASSES = {
    "floor", "table", "wall", "ceiling", "seat", "shelf", "counter", "unknown",
}


@dataclass
class Surface:
    """One semantic planar surface the user could place content on/near."""
    cls: str                              # one of SURFACE_CLASSES
    center_m: tuple = (0.0, 0.0, 0.0)     # world centre (x, y, z)
    size_m: tuple = (1.0, 1.0)            # extent along the surface's local (u, v)
    normal: tuple = (0.0, 1.0, 0.0)       # surface normal (up for floor/table)
    yaw_rad: float = 0.0                  # rotation of the extent rectangle about normal
    height_m: float = 0.0                 # height above floor (table≈0.74, floor=0)
    confidence: float = 1.0

    def as_dict(self):
        d = asdict(self)
        for k in ("center_m", "size_m", "normal"):
            d[k] = list(getattr(self, k))
        return d


@dataclass
class Obstacle:
    """A volume to keep content out of: furniture, clutter, a detected object."""
    bbox_min_m: tuple = (0.0, 0.0, 0.0)
    bbox_max_m: tuple = (0.0, 0.0, 0.0)
    category: str = "unknown"            # "table","chair","lamp",... (RoomPlan/CV)
    confidence: float = 1.0

    def as_dict(self):
        d = asdict(self)
        d["bbox_min_m"] = list(self.bbox_min_m)
        d["bbox_max_m"] = list(self.bbox_max_m)
        return d


@dataclass
class UserPose:
    position_m: tuple = (0.0, 0.0, 0.0)  # where the user stands
    forward: tuple = (0.0, 0.0, -1.0)    # gaze/heading (horizontal)
    eye_height_m: float = 1.45

    def as_dict(self):
        d = asdict(self)
        d["position_m"] = list(self.position_m)
        d["forward"] = list(self.forward)
        return d


@dataclass
class LightProfile:
    """Coarse lighting so content can be tinted/grounded to match the room."""
    intensity: float = 1000.0            # ARLightEstimate.ambientIntensity (lumens)
    color_temp_k: float = 6500.0         # ARLightEstimate.ambientColorTemperature
    key_direction: tuple = (0.0, -1.0, 0.0)  # best-effort primary light direction

    def as_dict(self):
        d = asdict(self)
        d["key_direction"] = list(self.key_direction)
        return d


@dataclass
class RoomModel:
    """The whole understood room. `free_floor_m` is a polygon (list of (x,z)) of
    walkable/placeable floor with clutter + a clearance margin removed."""
    surfaces: list = field(default_factory=list)        # list[Surface]
    obstacles: list = field(default_factory=list)        # list[Obstacle]
    free_floor_m: list = field(default_factory=list)     # [(x, z), ...] polygon
    user: UserPose = field(default_factory=UserPose)
    light: LightProfile = field(default_factory=LightProfile)
    bounds_m: tuple = (4.0, 3.0, 2.6)                    # (width, depth, ceiling)
    source: str = "default"                              # arkit|roomplan|cv|default|manual
    confidence: float = 1.0

    # -- convenience queries the solver uses -------------------------------------
    def surfaces_of(self, cls: str) -> list:
        return [s for s in self.surfaces if s.cls == cls]

    def best_table(self):
        tables = self.surfaces_of("table")
        return max(tables, key=lambda s: s.size_m[0] * s.size_m[1], default=None)

    def floor(self):
        floors = self.surfaces_of("floor")
        return max(floors, key=lambda s: s.size_m[0] * s.size_m[1], default=None)

    def as_dict(self):
        return {
            "schema": SCHEMA,
            "surfaces": [s.as_dict() for s in self.surfaces],
            "obstacles": [o.as_dict() for o in self.obstacles],
            "free_floor_m": [list(p) for p in self.free_floor_m],
            "user": self.user.as_dict(),
            "light": self.light.as_dict(),
            "bounds_m": list(self.bounds_m),
            "source": self.source,
            "confidence": self.confidence,
        }


# ── bridge to the existing coarse RoomProfile (analysis.py) ─────────────────────
# Lets the richer RoomModel flow into the current solver unchanged, and lets the
# old plane-only path keep working until the LiDAR producer lands.

def to_room_profile(rm: "RoomModel"):
    """Collapse a RoomModel down to the legacy RoomProfile analysis.estimate uses."""
    from analysis import RoomProfile  # studio/spatail on path
    floor = rm.floor()
    table = rm.best_table()
    fw = floor.size_m[0] if floor else rm.bounds_m[0]
    fd = floor.size_m[1] if floor else rm.bounds_m[1]
    return RoomProfile(
        floor_area_m2=float(fw * fd),
        floor_clear_w_m=float(fw), floor_clear_d_m=float(fd),
        table_present=table is not None,
        table_top_h_m=float(table.height_m) if table else 0.74,
        table_w_m=float(table.size_m[0]) if table else 1.2,
        table_d_m=float(table.size_m[1]) if table else 0.7,
        eye_height_m=float(rm.user.eye_height_m),
        source=rm.source,
    )


def from_room_profile(rp) -> "RoomModel":
    """Lift a coarse plane-only RoomProfile into a RoomModel (back-compat path)."""
    surfaces = [Surface(cls="floor", center_m=(0.0, 0.0, -1.5),
                        size_m=(rp.floor_clear_w_m, rp.floor_clear_d_m), height_m=0.0)]
    if rp.table_present:
        surfaces.append(Surface(cls="table", center_m=(0.0, rp.table_top_h_m, -0.74),
                                size_m=(rp.table_w_m, rp.table_d_m), height_m=rp.table_top_h_m))
    return RoomModel(
        surfaces=surfaces,
        user=UserPose(eye_height_m=rp.eye_height_m),
        bounds_m=(max(rp.floor_clear_w_m, 4.0), max(rp.floor_clear_d_m, 3.0), 2.6),
        source=rp.source,
    )


def demo() -> "RoomModel":
    """A representative living-room RoomModel for tests/authoring."""
    return RoomModel(
        surfaces=[
            Surface(cls="floor", center_m=(0.0, 0.0, -2.0), size_m=(4.0, 3.5), height_m=0.0),
            Surface(cls="table", center_m=(0.0, 0.74, -1.2), size_m=(1.2, 0.7), height_m=0.74),
            Surface(cls="wall", center_m=(0.0, 1.3, -3.6), size_m=(4.0, 2.6),
                    normal=(0.0, 0.0, 1.0)),
        ],
        obstacles=[Obstacle(bbox_min_m=(-1.8, 0.0, -2.8), bbox_max_m=(-1.0, 0.8, -2.0),
                            category="chair")],
        free_floor_m=[(-2.0, -0.4), (2.0, -0.4), (2.0, -3.4), (-2.0, -3.4)],
        user=UserPose(position_m=(0.0, 0.0, 0.0), forward=(0.0, 0.0, -1.0)),
        light=LightProfile(intensity=900.0, color_temp_k=4800.0),
        bounds_m=(4.0, 3.6, 2.6), source="demo",
    )


if __name__ == "__main__":
    import json
    rm = demo()
    print(json.dumps(rm.as_dict(), indent=2))
    print("floor:", rm.floor())
    print("best table:", rm.best_table())
